execution/function/function.go

```diff
@@ -0,0 +1,833 @@
+}
+
+func (o *functionOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	// Process only non-variadic functions.
+	if o.funcExpr.Func.Variadic == 0 {
+		return o.nextFunc(ctx)
+	}
+
+	return nil, nil
+}
+
+// Process single/multi-arg instant vector and scalar input functions.
+func (o *functionOperator) instantVectorOrScalarNext(ctx context.Context) ([]model.StepVector, error) {
+	// Call next on instant/range vector input.
+	input, err := o.nextOps[o.nextIndex].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	// Call next on all scalar inputs.
+	scalarInput := make([][]model.StepVector, len(o.nextOps)-1)
+	if len(scalarInput) != 0 {
+		s := 0
+		for i := range o.nextOps {
+			if i == o.nextIndex {
+				continue
+			}
+
+			si, err := o.nextOps[i].Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			scalarInput[s] = si
+			defer o.nextOps[i].GetPool().PutVectors(scalarInput[s])
+			s++
+		}
+	}
+
+	scalarPoints := make([]float64, len(scalarInput))
+	out := o.nextOps[o.nextIndex].GetPool().GetVectorBatch()
+	for v, vector := range input {
+		// scalar() depends on number of samples per vector and returns NaN if len(samples) != 1.
+		// So need to handle this separately here, instead of going via call which is per point.
+		if o.funcExpr.Func.Name == "scalar" {
+			if len(vector.Samples) <= 1 {
+				out = append(out, vector)
+				o.nextOps[o.nextIndex].GetPool().PutStepVector(vector)
+				continue
+			}
+
+			vector.Samples = vector.Samples[:1]
+			vector.SampleIDs = vector.SampleIDs[:1]
+			vector.Samples[0] = math.Float64frombits(value.NormalNaN)
+
+			out = append(out, vector)
+			o.nextOps[o.nextIndex].GetPool().PutStepVector(vector)
+			continue
+		}
+
+		for i := range vector.Samples {
+			// Scalar expressions can select from storage as well.
+			if len(o.funcExpr.Func.ArgTypes) > 1 && len(scalarInput) > 0 {
+				for l := range scalarInput {
+					if len(scalarInput[l]) >= 1 {
+						scalarPoints[l] = scalarInput[l][v].Samples[0]
+					}
+				}
+			}
+
+			// Call function by separately passing major input and scalars.
+			result := o.call(FunctionArgs{
+				Labels:       o.series[0],
+				Points:       []promql.Point{{V: vector.Samples[i]}},
+				StepTime:     vector.T,
+				ScalarPoints: scalarPoints,
+			})
+
+			vector.Samples[i] = result.V
+		}
+		out = append(out, vector)
+		o.nextOps[o.nextIndex].GetPool().PutStepVector(vector)
+	}
+	o.nextOps[o.nextIndex].GetPool().PutVectors(input)
+	return out, nil
+}
+
+// Process single arg range vector input functions.
+func (o *functionOperator) rangeVectorNext(ctx context.Context) ([]model.StepVector, error) {
+	input, err := o.nextOps[o.nextIndex].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	// We inject function call in tree, so only need to relay input from here.
+	return input, nil
+}
```

Comment:
Let's revert the Matrix selector nesting under Call so that we can remove the indirection here

Reply:
Reverted.